File: code/python/letterCombinations.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        dic = {
            2:['a','b','c'],
            3:['d','e','f'],
            4:['g','h','i'],
            5:['j','k','l'],
            6:['m','n','o'],
            7:['p','q','r','s'],
            8:['t','u','v'],
            9:['w','x','y','z']
        }
        
        if len(digits) == 0:
            return []
        if len(digits) == 1:
            return dic[int(digits)]
        
        A = []
        for i in range(1,len(digits)):
            cur = dic[int(digits[i])]
            if A == []:
                prev = dic[int(digits[i-1])]
            lenp = len(prev)
            lenc = len(cur)
            A = []
            for j in range(lenp):
                for k in range(lenc):
                    A.append(prev[j]+cur[k])
            
            prev = A
        
        return A
                

File: code/java/letterCombinations.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<String>();
        List<String> dummy = new ArrayList<String>();
        TreeMap map = new TreeMap();
        String temp;
        map.put('2',"abc");
        map.put('3',"def");
        map.put('4',"ghi");
        map.put('5',"jkl");        
        map.put('6',"mno");
        map.put('7',"pqrs");
        map.put('8',"tuv");
        map.put('9',"wxyz");
        if (digits.length() == 0)
            return result;
        if (digits.length() >= 1){
            temp = (String)map.get(digits.charAt(0));
            for(int i=0;i< temp.length();i++)
                result.add(temp.substring(i,i+1));
            if(digits.length() == 1)
                return result;
            
            for(int i=1;i<digits.length();i++){
                temp = (String)map.get(digits.charAt(i));               
                for(int j=0;j<result.size();j++)
                    for(int t=0;t<temp.length();t++)
                        dummy.add(((String)result.get(j))+temp.substring(t,t+1));
                        
                 result.clear();
                 for(int m=0;m<dummy.size();m++)
                     if(dummy.get(m).length() == i+1 )
                         result.add(dummy.get(m));
            }
        }        
        return result;
    }
}
